### ISSUE TITLE
Map becomes unpredictable after catching a NoSuchLayerException from MapboxMap.removeLayer

### ISSUE THREAD

**Issue by mattryavec** [mapbox/mapbox-gl-native#6507]
**Platform: Android**
**Mapbox SDK version: 4.2.0-beta3**

NOTE: This _might_ also happen after catching a NoSuchSourceException from removeSource.

This is definitely a minor issue, as the workaround is to simply **not** remove nonexistent layers or sources. But since `removeLayer` and `removeSource` explicitly declare `throws` for these exceptions, one would assume that it's appropriate to catch them and expect the map to behave normally.
### Steps to trigger behavior
1. Attempt to remove a nonexistent layer from the MapboxMap, but catch the NoSuchLayerException.
### Expected behavior

Map behaves normally
### Actual behavior

The map behavior is unpredictable. The worst thing that I found was when I attempted to move the MapView within the view hierarchy (i.e., remove it as the child of one ViewGroup and add it as the child of another). This works fine prior to caught NoSuchLayerException from removeLayer.

Other strange behavior was more subtle. It seemed to corrupt the native OpenGL rendering. For instance, if the MapView was resized, it generated GL_INVALID_VALUE errors. (Oddly enough, these got surfaced from another native lib we wrote that uses OpenGL and is pulling from glGetError).

**Comment by ivovandongen** [comment]
@mattryavec Thanks for the report! We'll look into it.

Could you maybe create a minimal case to reproduce this?

**Comment by ivovandongen** [comment]
> Could you maybe create a minimal case to reproduce this?

@mattryavec Any updates on this?

> The map behavior is unpredictable.

Could you explain what you mean with unpredictable?

**Comment by mattryavec** [comment]
Honestly, I've described it as best I could already. And like I said, there's a trivial workaround for it, so this report is more just an FYI for you all.

**Comment by ivovandongen** [comment]
@mattryavec Thanks, but let me rephrase; is the unexpected behavior something visually off? Or do you get crashes?

**Comment by mattryavec** [comment]
Ah, my mistake. Yes, there is a crash if you attempt to change the parent view of the MapView.

**Comment by ivovandongen** [comment]
> Yes, there is a crash if you attempt to change the parent view of the MapView.

@mattryavec Could you provide the relevant log and a snippet of code to see what you're doing? I've been playing around, but haven't been able to reproduce yet.

**Comment by hso-icd** [comment]
I am experiencing the same with 4.2.0-beta.4.

I call removeLayer(layerId) on a MapboxMap object with a non-existing layerId as parameter and catch a NoSuchLayerException. The code where this call is opens a different fragment. When I return to the fragment with the map, I get a crash. I have included the console output below.

If I use a variable to keep track of when the layer has been added and only remove it when it exists, everything works fine.

D/mbgl: {com.stay}[JNI]: nativeDestroySurface
D/mbgl: {com.stay}[Android]: NativeMapView::destroySurface
D/mbgl: {com.stay}[JNI]: nativeCreateSurface
D/mbgl: {com.stay}[Android]: NativeMapView::createSurface
D/mbgl: {com.stay}[JNI]: nativeFramebufferResize
W/Adreno-EGL: <qeglDrvAPI_eglSwapBuffers:3720>: EGL_BAD_SURFACE
E/mbgl: {com.stay}[OpenGL]: eglSwapBuffers() returned error 12301
D/AndroidRuntime: Shutting down VM
W/com.stay: type=1400 audit(0.0:11940): avc: denied { read } for name="mem" dev="debugfs" ino=1043964 scontext=u:r:untrusted_app:s0:c512,c768 tcontext=u:object_r:debugfs:s0 tclass=file permissive=0
W/com.stay: type=1400 audit(0.0:11941): avc: denied { read } for name="mem" dev="debugfs" ino=1043964 scontext=u:r:untrusted_app:s0:c512,c768 tcontext=u:object_r:debugfs:s0 tclass=file permissive=0
E/AndroidRuntime: FATAL EXCEPTION: main
                  Process: com.stay, PID: 11017
                  java.lang.Error: eglSwapBuffers() failed
                      at com.mapbox.mapboxsdk.maps.NativeMapView.nativeRender(Native Method)
                      at com.mapbox.mapboxsdk.maps.NativeMapView.render(NativeMapView.java:143)
                      at com.mapbox.mapboxsdk.maps.MapView.onDraw(MapView.java:1460)
                      at android.view.View.draw(View.java:16184)
                      at android.view.View.updateDisplayListIfDirty(View.java:15180)
                      at android.view.View.draw(View.java:15954)
                      at android.view.ViewGroup.drawChild(ViewGroup.java:3609)
                      at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3399)
                      at android.view.View.updateDisplayListIfDirty(View.java:15175)
                      at android.view.View.draw(View.java:15954)
                      at android.view.ViewGroup.drawChild(ViewGroup.java:3609)
                      at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3399)
                      at android.view.View.updateDisplayListIfDirty(View.java:15175)
                      at android.view.View.draw(View.java:15954)
                      at android.view.ViewGroup.drawChild(ViewGroup.java:3609)
                      at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3399)
                      at android.view.View.updateDisplayListIfDirty(View.java:15175)
                      at android.view.View.draw(View.java:15954)
                      at android.view.ViewGroup.drawChild(ViewGroup.java:3609)
                      at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3399)
                      at android.view.View.draw(View.java:16187)
                      at android.support.v4.view.ViewPager.draw(ViewPager.java:2341)
                      at android.view.View.updateDisplayListIfDirty(View.java:15180)
                      at android.view.View.draw(View.java:15954)
                      at android.view.ViewGroup.drawChild(ViewGroup.java:3609)
                      at android.support.design.widget.CoordinatorLayout.drawChild(CoordinatorLayout.java:1131)
                      at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3399)
                      at android.view.View.updateDisplayListIfDirty(View.java:15175)
                      at android.view.View.draw(View.java:15954)
                      at android.view.ViewGroup.drawChild(ViewGroup.java:3609)
                      at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3399)
                      at android.view.View.updateDisplayListIfDirty(View.java:15175)
                      at android.view.View.draw(View.java:15954)
                      at android.view.ViewGroup.drawChild(ViewGroup.java:3609)
                      at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3399)
                      at android.view.View.updateDisplayListIfDirty(View.java:15175)
                      at android.view.View.draw(View.java:15954)
                      at android.view.ViewGroup.drawChild(ViewGroup.java:3609)
                      at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3399)
                      at android.view.View.updateDisplayListIfDirty(View.java:15175)
                      at android.view.View.draw(View.java:15954)
                      at android.view.ViewGroup.drawChild(ViewGroup.java:3609)
                      at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3399)
                      at android.view.View.updateDisplayListIfDirty(View.java:15175)
                      at android.view.View.draw(View.java:15954)
                      at android.view.ViewGroup.drawChild(ViewGroup.java:3609)
                      at android.view.ViewGroup.dispatchDraw(ViewGroup.java:3399)
                      at android.view.View.draw(View.java:16187)
                      at com.android.internal.policy.PhoneWindow$DecorView.draw(PhoneWindow.java:2690)
                      at android.view.View.updateDisplayListIfDirty(View.java:15180)
                      at android.view.ThreadedRenderer.updateViewTreeDisplayList(ThreadedRenderer.java:281)
                      at android.view.ThreadedRenderer.updateRootDisplayList(ThreadedRenderer.java:287)
                      at android.view.ThreadedRenderer.draw(ThreadedRenderer.java:322)
                      at android.view.ViewRootImpl.draw(ViewRootImpl.java:2615)
                      at android.view.ViewRootImpl.performDraw(ViewRootImpl.java:2434)
                      at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2067)
                      at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1107)
                      at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:6013)
                      at android.view.Choreographer$CallbackRecord.run(Choreographer.java:858)
                      at android.view.Choreographer.doCallbacks(Choreographer.java:670)
                      at android.view.Choreographer.doFrame(Choreographer.java:606)
                      at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:844)
                      at android.os.Handler.handleCallback(Handler.java:739)
                      at android.os.Handler.dispatchMessage(Handler.java:95)
                      at android.os.Looper.loop(Looper.java:148)
                  	at android.app.ActivityThread.m

**Comment by jfirebaugh** [comment]
#7962 will have fixed this issue, since `NoSuchLayerException` and `NoSuchSourceException` are no longer thrown.
